File: subprojects/docs/src/snippets/plugins/buildService/groovy/buildSrc/src/main/java/DownloadPlugin.java
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.provider.Provider;

public class DownloadPlugin implements Plugin<Project> {
    public void apply(Project project) {
        // Register the service
        Provider<WebServer> serviceProvider = project.getGradle().getSharedServices().registerIfAbsent("web", WebServer.class, spec -> {
            // Provide some parameters
            spec.getParameters().getPort().set(5005);
        });

        project.getTasks().register("download", Download.class, task -> {
            // Connect the service provider to the task
            task.getServer().set(serviceProvider);
            // Declare the association between the task and the service
            task.usesService(serviceProvider);
            task.getOutputFile().set(project.getLayout().getBuildDirectory().file("result.zip"));
        });
    }
}

File: subprojects/docs/src/snippets/plugins/buildService/kotlin/buildSrc/src/main/kotlin/DownloadPlugin.kt
import org.gradle.api.Project
import org.gradle.api.Plugin
import org.gradle.api.provider.Provider

class DownloadPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        // Register the service
        val serviceProvider = project.gradle.sharedServices.registerIfAbsent("web", WebServer::class.java) {
            // Provide some parameters
            it.parameters.port.set(5005)
        }

        project.tasks.register("download", Download::class.java) {
            // Connect the service provider to the task
            it.server.set(serviceProvider)
            it.outputFile.set(project.layout.buildDirectory.file("result.zip"))
        }
    }
}
